mlflow/utils/search_utils.py

```diff
@@ -1,45 +1,198 @@
+        if token.ttype == TokenType.Operator.Comparison:
+            return {"comparator": token.value}
+        elif token.ttype in cls.VALUE_TYPES:
+            return {"value": token.value}
+        else:
+            return {}
+
+    @classmethod
+    def _get_comparison(cls, comparison):
+        comp = {}
+        for t in comparison.tokens:
+            if isinstance(t, Identifier):
+                comp.update(cls._get_identifier(t.value))
+            elif isinstance(t, Token):
+                comp.update(cls._process_token(t))
+        return comp
+
+    @classmethod
+    def _invalid_statement_token(cls, token):
+        if isinstance(token, Comparison):
+            return False
+        elif token.is_whitespace:
+            return False
+        elif token.match(ttype=TokenType.Keyword, values=["AND"]):
+            return False
+        else:
+            return True
+
+    @classmethod
+    def _process_statement(cls, statement):
+        # check validity
+        invalids = list(filter(cls._invalid_statement_token, statement.tokens))
+        if len(invalids) > 0:
+            invalid_clauses = ", ".join("'%s'" % token for token in invalids)
+            raise MlflowException("Invalid clause(s) in filter string: %s" % invalid_clauses)
+        return [cls._get_comparison(si) for si in statement.tokens if isinstance(si, Comparison)]
+
+    @classmethod
+    def search_expression_to_dict(cls, search_expression):
+        key_type = search_expression.WhichOneof('expression')
+        if key_type == cls._METRIC_IDENTIFIER:
+            key = search_expression.metric.key
+            metric_type = search_expression.metric.WhichOneof('clause')
+            if metric_type == 'float':
+                comparator = search_expression.metric.float.comparator
+                value = search_expression.metric.float.value
+            elif metric_type == 'double':
+                comparator = search_expression.metric.double.comparator
+                value = search_expression.metric.double.value
+            else:
+                raise MlflowException("Invalid metric type: '%s', expected float or double")
+            return {
+                "type": cls._METRIC_IDENTIFIER,
+                "key": key,
+                "comparator": comparator,
+                "value": value
+            }
+        elif key_type == cls._PARAM_IDENTIFIER:
+            key = search_expression.parameter.key
+            comparator = search_expression.parameter.string.comparator
+            value = search_expression.parameter.string.value
+            return {
+                "type": cls._PARAM_IDENTIFIER,
+                "key": key,
+                "comparator": comparator,
+                "value": value
+            }
+        else:
+            raise MlflowException("Invalid search expression type '%s'" % key_type)
+
+    def _parse(self):
+        if self._filter_string:
+            parsed = sqlparse.parse(self._filter_string)
```

Comment:
The docs for sqlparse ([link](https://sqlparse.readthedocs.io/en/latest/api/#sqlparse.parse)) aren't super explicit about whether this can throw an exception - it might be safe to just assume this never throws, but to be defensive we could try-catch here & raise an MlflowException with an INTERNAL_ERROR & the sqlparse exception message (otherwise I think the client will see a generic Flask-generated error message like):

```
2019/03/01 18:18:21 ERROR mlflow.utils.rest_utils: API request to http://localhost:5000/api/2.0/preview/mlflow/runs/search failed with code 500 != 200, retrying up to 0 more times. API response body: <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<title>500 Internal Server Error</title>
<h1>Internal Server Error</h1>
<p>The server encountered an internal error and was unable to complete your request.  Either the server is overloaded or there is an error in the application.</p>
```

Reply:
Added the extra check--but `sqlparse` is super robust and can parse any string. None string objects are dealt with due to type checking in protos